Werkgroep API design rules/naamgeving.md

```diff
@@ -0,0 +1,56 @@
+- Gebruik geen samengestelde namen voor geneste objecten (bv /adres_straat).
+    - Gebruik hiervoor een sub-object 'adres' met daarin een 'straat' parameter.
+
+## Resource bodies
+- Gebruik JSON of JSON gebaseerde documenttypen (bv JSON+HAL)
+- Gebruik camelCase elementen
+
+## Request parameters
+Het voorstel voor request parameters (query parameters en in de request body) is:
+
+- Technische request parameters (gebruikt om te filteren, sorteren, etc) bij voorkeur Engelstalig (`sort`, `filter`, `expand` etc);
+- GEEN gebruik van underscores (dus `sort` ipv `_sort`), omdat het anders zal betekenen dat (bijna) alle request parameters met een underscore worden geschreven.
+    - Zeker wanneer complexere vormen van filtering worden toegestaan, bijvoorbeeld `/collectie?filter[status]=open` in plaats van `/collectie?status=open`. 
+    - Tevens zullen conflicten tussen *technische* (bv. `sort`) en functionele (bv `status` in bovenstaand voorbeeld) beperkt blijven omdat technische request parameters Engelstalig zijn en functionele over het algemeen Nederlandstalig.
+- Verdere uitwerking van het gebruik van de technische request parameters voor filteren en sorteren is onderdeel van de werkgroep Filters en Sorteringen.
+
+## Enumeraties
+Het voorstel voor enumeraties is:
+
+- (MUST) Gebruik technische waarden in enumeraties die los staan van presentatie van deze waarden.
+- In het geval dat het domeinmodel presentatie-specifieke waarden bevat (bijvoorbeeld BAG objectstatusen), gebruik daarvoor bij voorkeur GEEN enumeraties maar strings; documenteer wel de mogelijke waarden goed in de API documentatie.
+    - Nadelen van het gebruik van presentatie-specifieke waarden als enums zijn:
+        - het is voor Clients lastig om fouten te maken wanneer een waarde moet worden opgegeven of aangepast middels een `POST`, `PUT` of `PATCH` operatie.
+        - wanneer een enum waarde moet worden aangepast omwille de behoefte om de presentatie te wijzigen, betekent dit een breaking API change.
+        - in het geval van meertaligheid moeten API Clients verschillende logica hanteren om enum waarden uit te lezen of te wijzigen, afhankelijk van de geselecteerde taal.
+    - Beter is om naast de presentatie-specifieke waarde ook een enum aan te bieden en de presentatie-specifieke waarde op basis van de (technische) enum waarde te laten berekenen door de API.
+        - Als voorbeeld voor de BAG object statusen kan bijvoorbeeld worden gekozen voor een veld pand.status.naam, welke een string is met waarden als `Bouw gestart` en een veld pand.status.code, welke een enum is met waarden als `BOUW_GESTART`. Wanneer de enum door de Client wordt gewijzigd, berekent de API een nieuwe waarde voor de `pand.status.naam`. Dit maakt het tevens gemakkelijk om meertaligheid toe te passen, waarbij `pand.status.naam` afhankelijk van de geselecteerde taal wordt gepresenteerd.
+- (SHOULD) Gebruik waar mogelijk UPPER_CAMEL_CASE, URI's voor enumeraties.
```

Comment:
is dat niet een beetje strijdig met elkaar? Uri's te schrijven in camelCase? Op andere plekken in de API zeggen we juist dat Uri's (zoals resource namen) lowercase (kebab-case) zijn.

Ik wil ervoor pleiten dat enumeratiewaarden ook betekenisvol moeten zijn. Bijvoorbeeld voor geslacht niet de waarde 1 (dus geen integer) of "M" (code), maar "MAN". Voordeel is dat de client developer in zijn code kan herkennen wat de enumeratiewaarde betekent. Dat kan fouten voorkomen.

Reply:
Hiermee werd bedoeld: termen in UPPER_CAMEL_CASE óf URIs.